src/store/middleware/api-request.js

```diff
@@ -0,0 +1,29 @@
+/* eslint-disable consistent-return */
+import axios from 'axios';
+import API_BASE_URL from '../../settings/index';
+
+const apiRequestMiddleware = store => next => async (action) => {
+  const { dispatch } = store;
+  // i.e. redux form actions
+  if (action.type !== 'API_REQUEST') {
+    return next(action);
+  }
+
+  const {
+    SUCCESS, FAILURE, endpoint, method, data,
+  } = action.payload;
+  const url = `${API_BASE_URL}/${endpoint}`;
+
+  try {
+    const response = await axios({
+      url,
+      method,
+      data,
+    });
+    dispatch({ type: SUCCESS, payload: response.data });
+  } catch (err) {
+    dispatch({ type: FAILURE, payload: err.response.data });
```

Comment:
The `catch` block here is only handling error cases where axios returns `err.response.data`. You have to consider the case where an error is thrown and it does not have the properties `response.data`. Otherwise, the `catch` block will throw an error trying to read `response.data` and your application will crash.

Separately, the structure of the payload in the error case is different from the expected payload in the success case. You are not handling that different structure in the component. So, you want to relook at what you're dispatching in the `catch` case.

Reply:
So are you proposing or suggesting that I should just do:

```
  } catch (err) {
    dispatch({ type: FAILURE, payload: err});
  }
```

And handle the `response.data` on the component that is expecting that property data?

I do not see what's wrong with having a different payload on error vs success. I've followed and read 2 tutorials and they did no mention anything about having same structures. Also... "You are not handling that different structure in the component. "... not sure what you mean, I am handling the different structures in my components. I.e. if you look at form, it renders an error message in the event there is an error and in the event there is a success, it re-directs to a different component.

Reply:
Let's breakdown the issues I've raised as they are separate:

1. In the `catch` block, you are dispatching `{ type: FAILURE, payload: err.response.data}`. My feedback here is that `err.response.data` does not always exist. So, in those cases you're going to get an error in your `catch` block and your application will crash.

2. Maybe I missed where you're handling this different structure in your component. But, I cannot find it. Please always include file and line number as I don't know where exactly you're talking about. (It helps to speed up my job as a reviewer :))

Reply:
1. I was just saying that I could update it so that the payload now includes the error object itself and not `error.response.data` and then I handle the response.data in the actual component that is consuming it. Separately I still don't think this is a issue since I'm only making API request to my back-end and I've designed my back-end to include in their response the response.data in the error. But I will update it nevertheless.
2. Eh... how I am handling the different structure is by checking the status code of the request in the payload. For example... in login/index.js line 34 `    const isLoginSuccessful = statusCode === 200;`. I check if the login was successful and then if it wasn't, I just render the formError component as shown in line 54 `{!isLoginSuccessful && this.renderFormError()}`. I still want to let the user interact with the form components so that they can try logging in again.

Reply:
1. Ye you can return the entire error object in the payload but that is an anti-pattern as you will end up handling the same structure in all components that get that response. Additionally, you have to treat the backend as if it is an external application which you do not have control over. The backend (API) and the frontend (client) should be designed based on a contract. And, the client should be designed to withstand changes to the contract in the backend. This is a similar problem we have why we created the monorepo at Lifion; the blocks code and the e2e-tests code had to have a co-ordinated deployment for every feature, which is a nightmare as we have seen.
2. Ah gotcha.. thank's for the reference. Looking at _src/components/user/login/index.js_ line 25, I'm seeing that if the error response does not have a status code, then "undefined" is going to be printed in the alert box, which is not intended. And, this can happen if anywhere in _src/store/middleware/api-request.js_ throws an error since that error would not be an API response error but instead a JavaScript error. This is why it is important to ensure that you are handle all error use cases in the catch block and only dispatching the relevant information in a format that your components can handle.

Reply:
Sorry I did not see your response til now. I did not get any notification for this. I thought you were busy enough to not reply, LOL.
1- "as you will end up handling the same structure in all components that get that response." which is why I did it how I had it originally from the beginning. I am out of options and googling a solution is not helping me. The only thing I can think of is just checking in src/store/middleware/api-request.js line 25 if that error has `response.data` and if so, send that. And if not, send the whole error object.
2- Undefined will not be printed. Please check line 27, there is a ternary which check for the status code. If it's undefined, it will return null. I had tested for such scenario.

Reply:
Thanks for the feedback! I have created the ticket and I have implemented #2. Gonna merge~